pkg/suggestion/v1beta1/pbt/service.py

```diff
@@ -0,0 +1,392 @@
+            return self._set_validate_context_error(
+                context, "Param(n_population) should be >= 5"
+            )
+        if not 0 <= float(settings["truncation_threshold"]) <= 1:
+            return self._set_validate_context_error(
+                context,
+                "Param(truncation_threshold) should be between 0 and 1, inclusive",
+            )
+        if (
+            "resample_probability" in settings
+            and not 0 <= settings["resample_probability"] <= 1
+        ):
+            return self._set_validate_context_error(
+                context,
+                "Param(resample_probability) should be null to perturb at 0.8 or 1.2, or be between 0 and 1, inclusive, to resample",
+            )
+
+        return api_pb2.ValidateAlgorithmSettingsReply()
+
+    def GetSuggestions(self, request, context):
+        if self.is_first_run:
+            settings = {
+                entry.name: entry.value
+                for entry in request.experiment.spec.algorithm.algorithm_settings
+            }  # all type:str
+            search_space = HyperParameterSearchSpace.convert(request.experiment)
+            search_space = [
+                HyperParameterSampler.from_hyperparameter(i)
+                for i in search_space.params
+            ]
+            # Always maximize the objective_metric
+            objective_scale = (
+                1 if request.experiment.spec.objective.type == api_pb2.MAXIMIZE else -1
+            )
+            objective_metric = request.experiment.spec.objective.objective_metric_name
+            # Create Job Manager
+            self.job_queue = PbtJobQueue(
+                int(settings["n_population"]),
+                float(settings["truncation_threshold"]),
+                None
+                if not "resample_probability" in settings
+                else float(settings["resample_probability"]),
+                search_space,
+                objective_metric,
+                objective_scale,
+            )
+
+            self.is_first_run = False
+
+        # Update states for completed trials
+        for trial in request.trials:
+            self.job_queue.update(trial)
+
+        # Synchronize generation
+        request_count = request.current_request_number
+        if len(self.job_queue) < request_count:
+            if len(self.job_queue) > 0:
+                logger.info(
+                    "Job queue < request count; flushing queue before spawning..."
+                )
+                request_count = len(self.job_queue)
+            elif len(self.job_queue.running) > 0:
+                logger.info(
+                    "Waiting for trials to complete before spawning next generation: {}".format(
+                        list(self.job_queue.running.keys())
+                    )
+                )
+                self.job_queue.verify_running(request.trials)
+                return api_pb2.GetSuggestionsReply(parameter_assignments=[])
```

Comment:
Won't we get an error here when [assignments != current request number](https://github.com/kubeflow/katib/blob/280eaed8cc27e25241fdcd0517354ed5deb732d9/pkg/controller.v1beta1/suggestion/suggestionclient/suggestionclient.go#L125-L128) ?
Or this use-case should not be possible ?

Reply:
It will error, but the client keeps retrying with increasing delay. This in the worst case will cause a job to pause for an extended duration while the retry timeout runs down. I experimented with changing `!=` to `>` in an earlier iteration, but wasn't sure if that would break other suggestion service assumptions (it works fine for a few E2E tests I manually went through, but breaks many unit tests; a few I wasn't able to easily track down).

Reply:
I see, thank you for explanation.
Are we sure that Suggestion controller will reconcile it again ?
As I can see, when [`ReconcileSuggestion`](https://github.com/kubeflow/katib/blob/master/pkg/controller.v1beta1/suggestion/suggestion_controller.go#L166-L177) failed, we don't requeue controller.
@gaocegege @johnugeorge @tenzen-y What do you think about this approach ?

Do we want to support "synchronous" algorithm service ? `GetSuggestions` (or any other API) is called only when all Trials have been completed in one generation ?

Reply:
My observation was that suggestions are retrieved whenever running jobs < parallel trials. I changed the implementation to async so we are no longer relying on the controller to keep polling until it gets a response of the appropriate size.